// @poppinss/oauth-client/build/src/UrlBuilder/index.js
"use strict";
/*
 * @poppinss/oauth-client
 *
 * (c) Harminder Virk <virk@adonisjs.com>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.UrlBuilder = void 0;
const querystring_1 = __importDefault(require("querystring"));
/**
 * Builds the url with query string
 */
class UrlBuilder {
    constructor(baseUrl) {
        this.baseUrl = baseUrl;
        this.params = {};
    }
    /**
     * Define the request param
     */
    param(key, value) {
        this.params[key] = value;
        return this;
    }
    /**
     * Clear a specific param
     */
    clearParam(key) {
        delete this.params[key];
        return this;
    }
    /**
     * Clear all params
     */
    clear() {
        this.params = {};
        return this;
    }
    /**
     * Returns the url
     */
    makeUrl() {
        const qs = querystring_1.default.stringify(this.params);
        return `${this.baseUrl.replace(/\/$/, '')}${qs ? `?${qs}` : ''}`;
    }
}
exports.UrlBuilder = UrlBuilder;
